File: src/main/java/net/treset/minecraftlauncher/launching/GameLauncher.java
package net.treset.minecraftlauncher.launching;

import javafx.util.Pair;
import net.hycrafthd.minecraft_authenticator.login.User;
import net.treset.mc_version_loader.launcher.LauncherInstanceDetails;
import net.treset.mc_version_loader.launcher.LauncherManifest;
import net.treset.minecraftlauncher.data.InstanceData;
import net.treset.minecraftlauncher.data.LauncherFiles;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

import java.io.IOException;
import java.util.List;
import java.util.Objects;
import java.util.function.Consumer;

public class GameLauncher {
    private static final Logger LOGGER = LogManager.getLogger(GameLauncher.class);

    private Pair<LauncherManifest, LauncherInstanceDetails> instance;
    private LauncherFiles files;
    private User minecraftUser;
    private List<Consumer<String>> exitCallbacks;
    private ResourceManager resourceManager;
    private GameListener gameListener;

    public GameLauncher(Pair<LauncherManifest, LauncherInstanceDetails> instance, LauncherFiles files, User minecraftUser, List<Consumer<String>> exitCallbacks) {
        this.instance = instance;
        this.files = files;
        this.minecraftUser = minecraftUser;
        this.exitCallbacks = exitCallbacks;
    }

    public boolean launch() {
        if(!files.isValid() || !files.reloadAll()) {
            LOGGER.warn("Unable to launch game: file reload failed");
            return false;
        }

        if(files.getLauncherDetails().getActiveInstance() != null && !files.getLauncherDetails().getActiveInstance().isBlank() && !cleanUpOldInstance()) {
            LOGGER.warn("Unable to launch game: unable to clean up old instance");
            return false;
        }

        InstanceData instanceData = InstanceData.of(instance, files);
        if(instanceData == null) {
            LOGGER.warn("Unable to launch game: instance data loaded incorrectly");
            return false;
        }
        resourceManager = new ResourceManager(instanceData);


        files.getLauncherDetails().setActiveInstance(instance.getKey().getId());
        if(!files.getLauncherDetails().writeToFile(files.getMainManifest().getDirectory() + files.getMainManifest().getDetails())) {
            LOGGER.warn("Unable to launch game: unable to write launcher details");
            return false;
        }

        if(!resourceManager.prepareResources()) {
            LOGGER.warn("Unable to launch game: unable to prepare resources");
            abortLaunch();
            return false;
        }

        if(!resourceManager.setLastPlayedTime()) {
            LOGGER.warn("Unable to launch game: unable to set last played time");
            abortLaunch();
            return false;
        }

        ProcessBuilder pb = new ProcessBuilder().redirectOutput(ProcessBuilder.Redirect.PIPE);
        CommandBuilder commandBuilder = new CommandBuilder(pb, instanceData, minecraftUser);
        if(!commandBuilder.makeStartCommand()) {
            LOGGER.warn("Unable to launch game: unable to set start command");
            abortLaunch();
            return false;
        }

        LOGGER.info("Starting game");
        LOGGER.debug("command=" + pb.command());
        try {
            Process p = pb.start();
            gameListener = new GameListener(p, resourceManager, exitCallbacks);
            gameListener.start();
            return true;
        } catch (IOException e) {
            LOGGER.warn("Unable to launch game: unable to execute command", e);
            abortLaunch();
            return false;
        }
    }

    private boolean cleanUpOldInstance() {
        LOGGER.debug("Cleaning up old instance resources: id=" + files.getLauncherDetails().getActiveInstance());
        boolean found = false;
        for(Pair<LauncherManifest, LauncherInstanceDetails> i : files.getInstanceComponents()) {
            if(Objects.equals(i.getKey().getId(), files.getLauncherDetails().getActiveInstance())) {
                InstanceData instanceData = InstanceData.of(i, files);
                if(instanceData == null) {
                    LOGGER.warn("Unable to cleanup old instance: instance data loaded incorrectly");
                    return false;
                }
                ResourceManager resourceManager = new ResourceManager(instanceData);

                if(!resourceManager.cleanupGameFiles()) {
                    LOGGER.warn("Unable to cleanup old instance: unable to cleanup resources");
                    return false;
                }
                found = true;
                break;
            }
        }
        if(!found) {
            LOGGER.warn("Unable to cleanup old instance: instance not found");
            return false;
        }
        LOGGER.info("Old instance resources cleaned up");
        return true;
    }

    private boolean abortLaunch() {
        LOGGER.warn("Aborting launch");
        if(!resourceManager.cleanupGameFiles()) {
            LOGGER.warn("Unable to abort launch correctly: failed to cleanup game files");
            return false;
        }
        files.getLauncherDetails().setActiveInstance(null);
        if(!files.getLauncherDetails().writeToFile(files.getMainManifest().getDirectory() + files.getMainManifest().getDetails())) {
            LOGGER.warn("Unable to abort launch: unable to write launcher details");
            return false;
        }
        return true;
    }

    public Pair<LauncherManifest, LauncherInstanceDetails> getInstance() {
        return instance;
    }

    public void setInstance(Pair<LauncherManifest, LauncherInstanceDetails> instance) {
        this.instance = instance;
    }

    public LauncherFiles getFiles() {
        return files;
    }

    public void setFiles(LauncherFiles files) {
        this.files = files;
    }

    public User getMinecraftUser() {
        return minecraftUser;
    }

    public void setMinecraftUser(User minecraftUser) {
        this.minecraftUser = minecraftUser;
    }

    public ResourceManager getResourceManager() {
        return resourceManager;
    }

    public GameListener getGameListener() {
        return gameListener;
    }

    public void setResourceManager(ResourceManager resourceManager) {
        this.resourceManager = resourceManager;
    }

    public void setGameListener(GameListener gameListener) {
        this.gameListener = gameListener;
    }

    public void setExitCallbacks(List<Consumer<String>> exitCallbacks) {
        this.exitCallbacks = exitCallbacks;
    }
}

File: client/composeApp/src/commonMain/kotlin/dev/treset/treelauncher/backend/launching/GameLauncher.kt
package dev.treset.treelauncher.backend.launching

import io.github.oshai.kotlinlogging.KotlinLogging
import dev.treset.mcdl.auth.data.UserData
import dev.treset.treelauncher.backend.data.LauncherFiles
import dev.treset.treelauncher.backend.data.manifest.InstanceComponent
import dev.treset.treelauncher.backend.util.QuickPlayData
import dev.treset.treelauncher.backend.util.exception.GameCommandException
import dev.treset.treelauncher.backend.util.exception.GameLaunchException
import dev.treset.treelauncher.backend.util.exception.GameResourceException
import java.io.IOException

class GameLauncher(
    val instance: InstanceComponent,
    val files: LauncherFiles,
    val offline: Boolean,
    val minecraftUser: UserData?,
    val quickPlayData: QuickPlayData? = null,
    var onExit: (String?) -> Unit = { _ -> },
    var onResourceCleanupFailed:  (Exception, (retry: Boolean) -> Unit) -> Unit  = { _, _ -> },
    var onExited: (String?) -> Unit = { _ -> },
) {
    private var resourceManager: ResourceManager? = null
    var gameListener: GameListener? = null

    @Throws(GameLaunchException::class)
    fun launch(cleanupActiveInstance: Boolean, doneCallback: (Exception?) -> Unit) {
        try {
            files.reload()
        } catch (e: IOException) {
            throw GameLaunchException("Unable to launch game: file reload failed", e)
        }
        if (!files.mainManifest.activeInstance.value.isNullOrBlank()) {
            if (!cleanupActiveInstance) {
                throw GameLaunchException("Unable to launch game: active instance already exists")
            }
            try {
                cleanUpOldInstance()
            } catch (e: GameLaunchException) {
                throw GameLaunchException("Unable to launch game: unable to clean up old instance", e)
            }
        }
        resourceManager = ResourceManager(instance)
        Thread(Runnable {
            try {
                resourceManager!!.setLastPlayedTime()
                resourceManager!!.prepareResources()
            } catch (e: GameResourceException) {
                val gle = GameLaunchException("Unable to launch game: unable to prepare resources", e)
                try {
                    abortLaunch()
                } catch (ex: GameLaunchException) {
                    doneCallback(GameLaunchException("Unable to launch game: unable to cleanup after fail", gle))
                }
                doneCallback(gle)
                return@Runnable
            } catch (e: IOException) {
                val gle = GameLaunchException("Unable to launch game: unable to prepare resources", e)
                try {
                    abortLaunch()
                } catch (ex: GameLaunchException) {
                    doneCallback(GameLaunchException("Unable to launch game: unable to cleanup after fail", gle))
                }
                doneCallback(gle)
                return@Runnable
            }
            try {
                finishLaunch()
                doneCallback(null)
            } catch (e: GameLaunchException) {
                doneCallback(e)
            }
        }).start()
    }

    fun stop() {
        gameListener?.stop()
    }

    @Throws(GameLaunchException::class)
    private fun finishLaunch() {
        val pb = ProcessBuilder().redirectOutput(ProcessBuilder.Redirect.PIPE)
        pb.inheritIO()
        val commandBuilder = CommandBuilder(pb, instance, offline, minecraftUser, quickPlayData)
        try {
            commandBuilder.makeStartCommand()
        } catch (e: GameCommandException) {
            abortLaunch()
            throw GameLaunchException("Unable to launch game: unable to set start command", e)
        }
        LOGGER.info { "Starting game" }
        try {
            val p = pb.start()
            resourceManager?.let { _ ->
                GameListener(p, this::cleanupResources).let {
                    gameListener = it
                    it.start()
                }
            }?: throw GameLaunchException("Unable to launch game: unable to create game listener: no resource manager")
        } catch (e: IOException) {
            abortLaunch()
            throw GameLaunchException("Unable to launch game: unable to execute command", e)
        }
    }

    @Throws(GameLaunchException::class)
    private fun cleanUpOldInstance() {
        LOGGER.debug { "Cleaning up old instance resources: id=${files.mainManifest.activeInstance}" }
        var found = false
        for (instance in files.instanceComponents) {
            if (instance.id == files.mainManifest.activeInstance) {
                val resourceManager = ResourceManager(instance)
                try {
                    resourceManager.cleanupResources()
                } catch (e: IOException) {
                    throw GameLaunchException("Unable to cleanup old instance: unable to cleanup resources", e)
                }

                found = true
                break
            }
        }
        if (!found) {
            throw GameLaunchException("Unable to cleanup old instance: instance not found")
        }
        LOGGER.info { "Old instance resources cleaned up" }
    }

    @Throws(GameLaunchException::class)
    private fun abortLaunch() {
        LOGGER.warn { "Aborting launch" }
        try {
            resourceManager?.cleanupResources() ?: throw GameLaunchException("Unable to abort launch correctly: no resource manager")
        } catch (e: IOException) {
            throw GameLaunchException("Unable to abort launch correctly: failed to cleanup game files")
        }
        files.mainManifest.activeInstance.value = null
        try {
            files.mainManifest.write()
        } catch (e: IOException) {
            throw GameLaunchException("Unable to abort launch correctly: failed to write launcher details")
        }
    }

    private fun cleanupResources(playDuration: Long, error: String?) {
        LOGGER.info { "Cleaning up resources" }
        onExit(error)
        try {
            resourceManager?.addPlayDuration(playDuration)
        } catch (e: IOException) {
            LOGGER.error(e) { "Unable to add play duration to statistics: duration=$playDuration" }
        }

        try {
            resourceManager?.cleanupResources()
            onExited(error)
        } catch (e: IOException) {
            onResourceCleanupFailed(e) { retry ->
                if(retry) {
                    cleanupResources(playDuration, error)
                } else {
                    onExited(error)
                }
            }
        }
    }

    companion object {
        private val LOGGER = KotlinLogging.logger {}
    }
}
